src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -143,10 +145,15 @@ class BlockFetcher(
           log.debug("Received {} block bodies that will be ignored", bodies.size)
           state.withBodiesFetchReceived
         } else {
-          log.debug("Fetched {} block bodies", bodies.size)
-          state.withBodiesFetchReceived.addBodies(peer, bodies)
+          state.validateBodies(bodies) match {
+            case Left(err) =>
+              peersClient ! BlacklistPeer(peer.id, err)
+              state.withBodiesFetchReceived
+            case Right(newBlocks) =>
+              log.debug("Fetched {} block bodies", newBlocks.size)
```

Comment:
I'd probably change things here a bit:
1. add logging about fetched bodies inconditionally
2. add logging about matching result 
3. change call to validate and then do sth into "addBodies", which validates bodies when adding them and results with new state, possible error and maybe information of how many bodies were matched with headers to blocks